src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CommandProcessor.java

```diff
@@ -1773,20 +1772,59 @@ public void doit(Tokens t) {
                 sysProps.run();
             }
         },
-        new Command("dumpheap", "dumpheap [filename]", false) {
+        new Command("dumpheap", "dumpheap [gz=<1-9>] [filename]", false) {
             public void doit(Tokens t) {
-                if (t.countTokens() > 1) {
+                int cntTokens = t.countTokens();
+                if (cntTokens > 2) {
                     usage();
                 } else {
                     JMap jmap = new JMap();
-                    String filename;
-                    if (t.countTokens() == 1) {
+                    String filename = "heap.bin";
+                    int gzlevel = 0;
+                    /*
+                     * Possible command:
+                     *     dumpheap gz=1 file;
+                     *     dumpheap gz=1;
+                     *     dumpheap file;
+                     *     dumpheap
+                     *
+                     * Use default filename if cntTokens == 0.
+                     * Handle cases with cntTokens == 1 or 2.
+                     */
+                    if (cntTokens > 2) {
+                        err.println("Too big number of options: " + cntTokens);
+                        usage();
+                        return;
+                    }
+                    if (cntTokens >= 1) { // parse first argument which is "gz=" option
+                        String option = t.nextToken();
+                        if (!option.startsWith("gz=")) {
+                            filename = option;
+                        } else {
+                            gzlevel = parseHeapDumpCompressionLevel(option);
+                            if (gzlevel <= 0 || gzlevel > 9) {
+                                usage();
+                                return;
+                            }
+                            filename = "heap.bin.gz";
+                        }
+                    }
+                    if (cntTokens == 2) { // parse second argument which is filename
+                        if (gzlevel == 0) {
+                            // The first option is not compression level, it is parsed as filename.
+                            err.println("unknown option: " + filename);
+                            usage();
+                            return;
+                        }
                         filename = t.nextToken();
-                    } else {
-                        filename = "heap.bin";;
+                        if (filename.startsWith("gz=")) {
+                            err.println("Filename should not start with \"gz=\": " + filename);
+                            usage();
+                            return;
+                        }
```

Comment:
I think having `if (cntTokens >= 1)` and `if (cntTokens == 2)` sections makes it harder to follow. Also, after calling parseHeapDumpCompressionLevel() you only need to check if 0 was returned. Otherwise you know it will be with the range of 1 to 9. I think the following might be easier to follow:
```
                    if (cntTokens == 1) { // first argument could be filename or "gz="
                        String option = t.nextToken();
                        if (!option.startsWith("gz=")) {
                            filename = option;
                        } else {
                            gzlevel = parseHeapDumpCompressionLevel(option);
                            if (gzlevel == 0) {
                                usage();
                                return;
                            }
                            filename = "heap.bin.gz";
                        }
                    }
                    if (cntTokens == 2) { // first argument is "gz=" followed by filename
                        gzlevel = parseHeapDumpCompressionLevel(option);
                        if (gzlevel == 0) {
                            usage();
                            return;
                        }
                        filename = t.nextToken();
                        if (filename.startsWith("gz=")) {
                            err.println("Filename should not start with \"gz=\": " + filename);
                            usage();
                            return;
                        }
                    }
```